File: code-jam/2022/1A/weightlifting/official_solution.py
# Test Set 1: OK
# Test Set 2: Time Limit Exceeded

nof_tests = int(input())
for test in range(1, nof_tests + 1):
    nof_exe, nof_weights = map(int, input().split())
    weigths_per_exe = [list(map(int, input().split())) for _ in range(nof_exe)]

    # Compute common sets of weights
    C = [[0] * nof_exe for _ in range(nof_exe)]
    for l in range(nof_exe):
        cs = list(weigths_per_exe[l])
        for r in range(l, nof_exe):
            for w in range(nof_weights):
                cs[w] = min(cs[w], weigths_per_exe[r][w])
                C[l][r] += cs[w]
    
    dp = [[float("+inf")] * nof_exe for _ in range(nof_exe)]
    for i in range(nof_exe):
        dp[i][i] = 0
    for size in range(1, nof_exe):
        for l in range(nof_exe - size):
            r = l + size
            for x in range(l, r):
                dp[l][r] = min(dp[l][r], dp[l][x] + 2 * (C[l][x] - C[l][r]) + dp[x + 1][r] + 2 * (C[x + 1][r] - C[l][r]))

    print("Case #{}: {}".format(test, dp[0][nof_exe - 1] + 2 * C[0][nof_exe - 1]))

File: code-jam/2022/1A/weightlifting/official_solution.cpp
#include <iostream>
#include <string.h>

using namespace std;

int MAX = 2e6;

int main(int argc, char *argv[]) {
    int t;
    cin >> t;
    for (int it = 1; it <= t; it++) {
        int e, w;
        cin >> e >> w;
        int x[e][w];
        for (int i = 0; i < e; i++)
            for (int j = 0; j < w; j++)
                cin >> x[i][j];

        // Compute common sets of weights
        int C[e][e] {};
        int set[w];
        for (int l = 0; l < e; l++) {
            memcpy(set, x[l], sizeof(x[l]));
            for (int r = l; r < e; r++)
                for (int iw = 0; iw < w; iw++) {
                    set[iw] = min(set[iw], x[r][iw]);
                    C[l][r] += set[iw];
                }
        }

        int dp[e][e];
        fill(*dp, *dp + e * e, MAX);
        for (int i = 0; i < e; i++)
            dp[i][i] = 0;
        for (int s = 1; s < e; s++)
            for (int l = 0; l < e - s; l ++) {
                int r = l + s;
                for (int x = l; x < r; x++) {
                    dp[l][r] = min(dp[l][r], dp[l][x] + 2 * (C[l][x] - C[l][r]) + dp[x + 1][r] + 2 * (C[x + 1][r] - C[l][r]));
                }
            }

        cout << "Case #" << it << ": " << dp[0][e - 1] + 2 * C[0][e - 1] << endl;
    }
    return 0;
}
